File: Keypad_4x4.py
import RPi.GPIO as GPIO
from time import sleep

class Keypad:
    def __init__(self):
        # GPIO pin configuration
        self.ROWS = [18, 23, 24, 25]    # GPIO pins for rows
        self.COLS = [4, 17, 27, 22]      # GPIO pins for columns

        # Keypad configuration (4x4)
        self.KEYPAD = [
            [1, 2, 3, 'A'],
            [4, 5, 6, 'B'],
            [7, 8, 9, 'C'],
            ['*', 0, '#', 'D']
        ]

        # Setup GPIO
        self.setup()

    def setup(self):
        GPIO.setwarnings(False)
        GPIO.setmode(GPIO.BCM)

        # Set column as outputs and set them to LOW
        for col in self.COLS:
            GPIO.setup(col, GPIO.OUT, initial=GPIO.LOW)

        # Set up row as inputs with pull-down resistors to avoid floating
        for row in self.ROWS:
            GPIO.setup(row, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)

    def read_keypad(self):
        for col_idx, col in enumerate(self.COLS):
            GPIO.output(col, GPIO.HIGH)  # set output to HIGH, one at a time
            for row_idx, row in enumerate(self.ROWS):
                if GPIO.input(row) == GPIO.HIGH:  # Check if the key is pressed
                    sleep(0.2) #Debounce time
                    key = self.KEYPAD[row_idx][col_idx]
                    while GPIO.input(row) == GPIO.HIGH:
                        pass
                    GPIO.output(row, GPIO.LOW)  # Deactivate output after the key pressed
                    return key
            GPIO.output(col, GPIO.LOW)  # Deactivate the output after scan the input
        return None

    def print_keypad(self):
        key = self.read_keypad()
        if key:
            print(f"Key Pressed: {key}")

    def cleanup(self):
        GPIO.cleanup()  # Clean up GPIO

if __name__ == "__main__":
    # Initialize the Keypad
    keypad = Keypad()

    try:
        while True:
            keypad.print_keypad()
    except KeyboardInterrupt:
        keypad.cleanup()  # Clean up GPIO on CTRL+C exit
        print("\nApplication stopped!")

File: Code in CPP/Keypad_4x4.cpp
#include <iostream>
#include <wiringPi.h>
#include <signal.h>
#include <chrono>

class Keypad {
private:
    // GPIO pin configuration (using wiringPi pin numbers)
    int ROWS[4] = {1, 4, 5, 6};    // wPi pin numbers for rows
    int COLS[4] = {7, 0, 2, 3};    // wPi pin numbers for columns

    // Keypad configuration (4x4)
    char KEYPAD[4][4] = {
        {'1', '2', '3', 'A'},
        {'4', '5', '6', 'B'},
        {'7', '8', '9', 'C'},
        {'*', '0', '#', 'D'}
    };

    // Debounce time in milliseconds
    const int DEBOUNCE_TIME = 50; // Adjust as needed

    // Last key state
    char lastKey = '\0';
    std::chrono::steady_clock::time_point lastPressTime;

    void cleanup() {
        // Set all pins to input with internal pull-down resistors
        for (int i = 0; i < 4; i++) {
            pinMode(ROWS[i], INPUT);
            pullUpDnControl(ROWS[i], PUD_DOWN); // Set as pull-down resistor
            pinMode(COLS[i], INPUT);
            pullUpDnControl(COLS[i], PUD_DOWN); // Set as pull-down resistor
        }
    }

public:
    Keypad() {
        wiringPiSetup(); // Initialize WiringPi

        // Configure row pins as input with internal pull-down resistors
        for (int i = 0; i < 4; i++) {
            pinMode(ROWS[i], INPUT);
            pullUpDnControl(ROWS[i], PUD_DOWN);
        }

        // Configure column pins as output and LOW
        for (int i = 0; i < 4; i++) {
            pinMode(COLS[i], OUTPUT);
            digitalWrite(COLS[i], LOW);
        }

        // Setup signal handler for ctrl+c
        signal(SIGINT, [](int sig) {
            Keypad::instance()->cleanup();
            exit(0);
        });

        lastKey = '\0';
        lastPressTime = std::chrono::steady_clock::now();
    }

    static Keypad* instance() {
        static Keypad instance;
        return &instance;
    }

    char getKey() {
        char key = '\0';
        for (int col = 0; col < 4; col++) {
            // Set current column to HIGH
            digitalWrite(COLS[col], HIGH);

            for (int row = 0; row < 4; row++) {
                // If the button at the selected row and column is pressed
                if (digitalRead(ROWS[row]) == HIGH) {
                    key = KEYPAD[row][col];
                    break;
                }
            }

            // Set column back to LOW after scanning
            digitalWrite(COLS[col], LOW);
            if (key != '\0') break;
        }

        // Handle debounce and key press duration
        auto now = std::chrono::steady_clock::now();
        auto elapsedTime = std::chrono::duration_cast<std::chrono::milliseconds>(now - lastPressTime).count();

        if (key != lastKey) {
            lastKey = key;
            lastPressTime = now;
        } else if (key != '\0' && elapsedTime > DEBOUNCE_TIME) {
            lastPressTime = now; // Reset debounce timer
        }

        return key;
    }

    void scan() {
        while (true) {
            char key = getKey();
            if (key != '\0') {
                std::cout << "Key Pressed: " << key << std::endl;
            }
            delay(10); // Small delay for CPU efficiency
        }
    }
};

int main() {
    Keypad* keypad = Keypad::instance();
    keypad->scan();
    return 0;
}
